src/OrbitGl/CallstackThreadBar.cpp

```diff
@@ -226,60 +226,117 @@ bool CallstackThreadBar::IsEmpty() const {
+  return module_and_function_name.module_is_unknown
+             ? absl::StrFormat("<i>%s</i>", module_and_function_name.module_name)
+             : module_and_function_name.module_name;
 }
 
-constexpr const char* kErrorColorString = "#ffb000";
+std::string CallstackThreadBar::FormatFunctionName(
+    const CallstackThreadBar::UnformattedModuleAndFunctionName& module_and_function_name,
+    int max_length) {
+  const std::string& function_name = module_and_function_name.function_name;
+  const std::string shortened_function_name =
+      (max_length > 0) ? ShortenStringWithEllipsis(function_name, static_cast<size_t>(max_length))
+                       : function_name;
+  // Simple HTML escaping.
+  const std::string escaped_function_name =
+      absl::StrReplaceAll(shortened_function_name, {{"&", "&amp;"}, {"<", "&lt;"}, {">", "&gt;"}});
+  return module_and_function_name.function_is_unknown
+             ? absl::StrFormat("<i>%s</i>", escaped_function_name)
+             : escaped_function_name;
+}
 
-std::string CallstackThreadBar::FormatCallstackForTooltip(const CallstackInfo& callstack,
-                                                          int max_line_length, int max_lines,
-                                                          int bottom_n_lines) const {
-  std::string result;
-  int size = static_cast<int>(callstack.frames().size());
-  if (max_lines <= 0) {
-    max_lines = size;
-  }
-  const int bottom_n = std::min(std::min(max_lines - 1, bottom_n_lines), size);
-  const int top_n = std::min(max_lines, size) - bottom_n;
+constexpr const char* kUnwindErrorColorString = "#ffb000";
+
+std::string CallstackThreadBar::FormatCallstackForTooltip(const CallstackInfo& callstack) const {
+  constexpr int kMaxLineLength = 100;
+  constexpr int kMaxLines = 20;
+  constexpr int kBottomLineCount = 5;
+
+  int callstack_size = static_cast<int>(callstack.frames().size());
+  const int bottom_n = std::min(std::min(kMaxLines - 1, kBottomLineCount), callstack_size);
+  const int top_n = std::min(kMaxLines, callstack_size) - bottom_n;
 
+  constexpr int kShortenedForReadabilityFakeIndex = -1;
+  constexpr const char* kShortenedForReadabilityString = "<i>... shortened for readability ...</i>";
+  std::vector<int> frame_indices_to_display;
   for (int i = 0; i < top_n; ++i) {
-    // The first frame is always correct.
-    if (callstack.IsUnwindingError() && i > 0) {
-      result.append(absl::StrFormat("<br/><span style=\" color:%s;\">%s</span>", kErrorColorString,
-                                    SafeGetFormattedFunctionName(callstack, i, max_line_length)));
-    } else {
-      result.append("<br/>" + SafeGetFormattedFunctionName(callstack, i, max_line_length));
-    }
+    frame_indices_to_display.push_back(i);
   }
-  if (max_lines < size) {
-    result += "<br/><i>... shortened for readability ...</i>";
+  if (kMaxLines < callstack_size) {
+    frame_indices_to_display.push_back(kShortenedForReadabilityFakeIndex);
```

Comment:
Nit: I would like to see an screenshot to see how it looks. But if you've already checked this, disregard this comment.

Reply:
Only if you really want one. I would have to find a long callstack. This functionality is unchanged.